lib/parser.js

```diff
@@ -145,43 +159,48 @@ function scope (scanner) {
 }
 
 /*
- * <body>          ::= <footer>+
- *                  | [<text>], <newline>, <body>*
- *                  | [<text>]
+ * <body>          ::= [<any text except pre-footer>], <newline>, <body>*
+ *                  | [<any text except pre-footer>]
  */
 function body (scanner) {
   const node = scanner.enter('body', [])
-  // 1*<footer>
-  while (!scanner.eof()) {
-    const f = footer(scanner)
-    if (f instanceof Error) {
-      scanner.abort(node)
-      node.children = []
-      // [<text>], <newline>, <body>*
-      const t = text(scanner)
-      // [<text>]
-      node.children.push(t)
-      // <newline>, <body>*
-      if (isNewline(scanner.peek())) {
-        scanner.next()
-        const b = body(scanner)
-        if (b instanceof Error) {
-          return b
-        } else {
-          Array.prototype.push.apply(node.children, b.children)
-        }
-      }
-      break // The recursive body(scanner) step should consume all tokens.
-    } else {
-      node.children.push(f)
+  // <any text except pre-footer>
+  const pf = preFooter(scanner)
+  if (!(pf instanceof Error)) return scanner.abort(node)
+
+  // [<text>]
+  const t = text(scanner)
+  node.children.push(t)
+  // <newline>, <body>*
+  if (isNewline(scanner.peek())) {
+    scanner.next()
+    const b = body(scanner)
+    if (!(b instanceof Error)) {
+      Array.prototype.push.apply(node.children, b.children)
     }
   }
   return scanner.exit(node)
 }
 
+/*
+ * <newline>*, <footer>+
+ */
+function preFooter (scanner) {
+  const node = scanner.enter('pre-footer', [])
+  while (isNewline(scanner.peek())) {
+    scanner.next()
+  }
+  let f
+  while (!scanner.eof()) {
```

Comment:
I'm not sure if I follow this code, won't this "consume" the `footer` without storing or rewinding it?

Reply:
Line `:168` rewinds the whole `body` parse if it hits a `preFooter`:

` if (!(pf instanceof Error)) return scanner.abort(node)`

This is meant to parse: `<any text except pre-footer>`.